### PR TITLE
Funcbench on GKE

### PR DESCRIPTION
**Pull request by nevill**
This is a replacement of #311 :)

You can check [my repository](https://github.com/nevill/prometheus/blob/master/.github/workflows/funcbench.yml) to see the corresponding workflow file.

Currently, the whole process takes over 70 minutes, but the automatically created token only can [last 60 minutes](https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#about-the-github_token-secret), so we have to use personal token here. In future, a better solution would be to send the benchmark result via commentMonitor.

@krasi-georgiev @geekodour Please review it.

### CODE REVIEW

File: funcbench/manifests/benchmark/2_job.yaml
Comment (geekodour) [inline funcbench/manifests/benchmark/2_job.yaml]: I think passing the full flag names for funcbench here will improve readability.

File: funcbench/manifests/benchmark/2_job.yaml
Comment (geekodour) [inline funcbench/manifests/benchmark/2_job.yaml]: I am removing the use of `GITHUB_SHA` in an upcoming PR as it is not of any use inside funcbench, we can either wait for it to get merged or merge this before.

**Review by geekodour — COMMENTED** [review]
I think the file location for the manifest files should be:

`funcbench/manifests/benchmark/1_namespace.yaml` -> `prombench/manifest/funcbench/benchmark/1_namespace.yaml` , similar for others.

and `make` targets can be added to `/prombench/Makefile` with appropriate funcbench specific target names.

File: funcbench/manifests/benchmark/2_job.yaml
Comment (geekodour) [reply funcbench/manifests/benchmark/2_job.yaml]: also please change the name of the image to  ` docker.io/prominfra/funcbench:master` from `nevill/funcbench:20200324`

Comment (nevill) [reply funcbench/manifests/benchmark/2_job.yaml]: Probably I need another PR to change the image name ? Otherwise you cannot really test my workflow. Because `prominfra/funcbench:master` is not sync with this PR. When `prombench` executes the manifest, it cannot find the appropriate docker image.

File: funcbench/manifests/benchmark/2_job.yaml
Comment (nevill) [reply funcbench/manifests/benchmark/2_job.yaml]: Either way is fine. Just to mention that, `GITHUB_SHA` is still used in [funcbench](https://github.com/prometheus/test-infra/blob/master/funcbench/env.go#L129) now.

**Comment by nevill** [comment]
> and `make` targets can be added to `/prombench/Makefile` with appropriate funcbench specific target names.

@geekodour You mean to add a new target like `make funcbench_resource_apply` here ?

**Comment by geekodour** [comment]
> You mean to add a new target like make funcbench_resource_apply here ?

yes. it'll simplify the github action workflow aswell.

File: funcbench/manifests/benchmark/2_job.yaml
Comment (geekodour) [reply funcbench/manifests/benchmark/2_job.yaml]: okay we keep it this way for testing now then, just a reminder to change it to the correct name later in the same PR before merging.

**Comment by nevill** [comment]
Updated Makefile and moved manifest files.

**Comment by geekodour** [comment]
@nevill can you update the README for funcbench in #343 with manual instructions to run funcbench on gke like this: 
https://github.com/prometheus/test-infra/tree/master/prombench#start-a-benchmarking-test-manually

File: funcbench/README.md
Comment (geekodour) [inline funcbench/README.md:69]: Would this not need deleting the funcbench namespace or deleting the nodepool afterwards?

Comment (nevill) [reply funcbench/README.md:69]: Added more steps below to recycle them.

**Comment by geekodour** [comment]
I think it'll be good to persist the job for sometime after funcbench finishes so that funcbench users are able to see the logs if needed. 

https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#ttl-mechanism-for-finished-jobs

currently gke jobs logs will probably be only available via stackdriver right?

File: funcbench/README.md
Comment (geekodour) [inline funcbench/README.md]: > Arguments for the benchcmp are read from files created by previous action

We're no longer doing this, we're getting the arguments from the `repository_dispatch` event.

File: prombench/Makefile
Comment (geekodour) [inline prombench/Makefile]: Should we change `BRANCH` to `TARGET` ?

File: prombench/manifests/funcbench/benchmark/2_job.yaml
Comment (geekodour) [inline prombench/manifests/funcbench/benchmark/2_job.yaml]: docker.io/prominfra/funcbench:master

**Review by geekodour — COMMENTED** [review]
Some minor changes.

**Comment by nevill** [comment]
> I think it'll be good to persist the job for sometime after funcbench finishes so that funcbench users are able to see the logs if needed.
> 
> https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#ttl-mechanism-for-finished-jobs
> 
> currently gke jobs logs will probably be only available via stackdriver right?

I discussed with @krasi-georgiev about it some time ago. Currently we just remove everything, including node pools, otherwise, we have to wait until job gets recycled, then remove node pools, which includes more steps. We can add a flag to set if we keep the job and node pool for debugging, but this can be a new feature.

File: prombench/Makefile
Comment (nevill) [reply prombench/Makefile]: Not sure about it. Some places (like, in `REAMD.md`) are still calling it `branch`. I think we use branch most of the time in the comments, let's keep it like this first.

Comment (geekodour) [inline prombench/Makefile]: we can get rid of `LAST_COMMIT_SHA` variable now.

File: funcbench/README.md
Comment (geekodour) [inline funcbench/README.md]: I don't see any second mention of `PERSONAL_TOKEN`. Is something missing?

File: funcbench/README.md
Comment (geekodour) [inline funcbench/README.md:35]: Just a reminder: Once we finalize on the workflow with GitHub statuses and action failure handling, I think linking this section to the workflow directly will be a good idea.

**Review by geekodour — COMMENTED** [review]
Would like @krasi-georgiev to take a final look.

File: funcbench/README.md
Comment (nevill) [reply funcbench/README.md]: No, it's only set in the workflow file.

File: funcbench/README.md
Comment (krasi-georgiev) [inline funcbench/README.md]: Why do we need this link?

File: funcbench/README.md
Comment (krasi-georgiev) [inline funcbench/README.md]:
```suggestion
        GITHUB_TOKEN: ${{ secrets.PERSONAL_TOKEN }} # The GH action token lasts up to 60min so using PERSONAL_TOKEN guarantees that can post back the results even when the bench tests takes longer.
```

File: funcbench/README.md
Comment (krasi-georgiev) [inline funcbench/README.md]: hm, why not provide the regex directly here? It will be easier to change and understand.

File: funcbench/README.md
Comment (krasi-georgiev) [inline funcbench/README.md]: is that nececary?

File: pkg/provider/k8s/k8s.go
Comment (krasi-georgiev) [inline pkg/provider/k8s/k8s.go]: Why is this delete necessary?

File: prombench/manifests/funcbench/benchmark/2_job.yaml
Comment (krasi-georgiev) [inline prombench/manifests/funcbench/benchmark/2_job.yaml]: can you remind me what is this arg and why is it needed?

**Review by krasi-georgiev — COMMENTED** [review]
Nice job, very easy to follow and understand!

just a few small questions and can merge it.

File: prombench/manifests/funcbench/benchmark/2_job.yaml
Comment (nevill) [reply prombench/manifests/funcbench/benchmark/2_job.yaml]: `workspace` is a replacement of env variable `GITHUB_WORKSPACE`, it's the directory where the source will be cloned to.

File: funcbench/README.md
Comment (geekodour) [reply funcbench/README.md]: I think this stayed after testing it with the prometheus-community repo for prombench, It can also be seen with the [prombench workflow file.](https://github.com/prometheus/prometheus/blob/386aea77744eac455c7430bd33cd0c736dba8a87/.github/workflows/prombench.yml#L28) keeping it here does make it easier to test though, as one will just need to copy and paste the workflow and just replace the variables.

Comment (geekodour) [reply funcbench/README.md]: Should we remove it? as it's set to the same link in the Dockerfile?

File: pkg/provider/k8s/k8s.go
Comment (nevill) [reply pkg/provider/k8s/k8s.go]: It's inherited from #311, at that time, we didn't have the step `make funcbench_resource_delete`. Now we can keep previous step and remove this line.

File: funcbench/README.md
Comment (nevill) [reply funcbench/README.md]: see the next comments.

File: funcbench/README.md
Comment (nevill) [reply funcbench/README.md]: I use it for testing my fork. In real case, we don't need to set it.

File: funcbench/README.md
Comment (nevill) [reply funcbench/README.md]: Since it's an example triggered by `repository_dispatch` event. ;)

File: funcbench/README.md
Comment (krasi-georgiev) [reply funcbench/README.md]: my question was really - what does this link specify? Why do we need to specify the test-infra repo? Does it download any files from there?

File: funcbench/README.md
Comment (krasi-georgiev) [reply funcbench/README.md]: I still don't understand why do we need to specify it elsewhere and not directly here?

Unless there is any real use case or a benefit from this don't see why it should be passed from another component. It defies the KISS principle :smile: 

@geekodour what do you think?

File: funcbench/README.md
Comment (krasi-georgiev) [reply funcbench/README.md]: sorry I am not sure which comments are you referring to here?

File: prombench/manifests/funcbench/benchmark/2_job.yaml
Comment (krasi-georgiev) [reply prombench/manifests/funcbench/benchmark/2_job.yaml]: can you give me an example of why do we need to have this as an argument and why can't we always use the GITHUB_WORKSPACE env?

**Comment by krasi-georgiev** [comment]
IIRC funcbnech can also be run locally right? If that is true it would be a good idea to also add a section on how to run locally. If you prefer to do this in a separate PR please open an issue so that we don't forget about it.

Having the examples of how to run locally without GH actions would also give some good context on the usage of some cli arguments.

**Comment by nevill** [comment]
> Having the examples of how to run locally without GH actions would also give some good context on the usage of some cli arguments.

Yes, it's [there](https://github.com/prometheus/test-infra/tree/master/funcbench#usage-examples) after #346 got merged.

File: prombench/manifests/funcbench/benchmark/2_job.yaml
Comment (nevill) [reply prombench/manifests/funcbench/benchmark/2_job.yaml:24]: It's a change introduced by another [pull request](https://github.com/prometheus/test-infra/pull/346#issue-401923405). In this PR, we update the example based on the change.

File: funcbench/README.md
Comment (nevill) [reply funcbench/README.md]: Basically, it will clone the TEST_INFRA_REPO to WORKSPACE, then run `prombench` against TEST_INFRA_REPO, in this case, we can update the manifests without rebuilding the docker image.
As a reference, this `TEST_INFRA_REPO` was introduced in #266.

File: funcbench/README.md
Comment (nevill) [reply funcbench/README.md]: `REGEX` is an input from user, we need comment-monitor to extract it out, then pass it to GH action via `repository_dispatch` event.
Do you suggest switching back to `issue_comment` event ?

File: funcbench/README.md
Comment (nevill) [reply funcbench/README.md]: I'm referring to [this](https://github.com/prometheus/test-infra/pull/343#discussion_r408710386). Line 186 `_ = c.jobDelete(resource)` is already removed from code.
So we have to remove the `Job` resource everytime it finishes, no matter it's successful or failed.

File: prombench/manifests/funcbench/benchmark/2_job.yaml
Comment (nevill) [reply prombench/manifests/funcbench/benchmark/2_job.yaml]: Or, are you suggesting `--workspace ${GITHUB_WORKSPACE}` ?

File: funcbench/README.md
Comment (krasi-georgiev) [reply funcbench/README.md]: yeah I remember now, don't think we should expose this in the workflow file as there is a very, very small chance someone would ever need to change this. Let's remove it from the example and leave it set to the default.

File: funcbench/README.md
Comment (krasi-georgiev) [reply funcbench/README.md]: no I was referring only to the `if: always()` line. Would this execute always even without this line?

File: prombench/manifests/funcbench/benchmark/2_job.yaml
Comment (krasi-georgiev) [reply prombench/manifests/funcbench/benchmark/2_job.yaml]: no I mean do you even need to set this? Why not have some default so clone to `/tmp` and that is it?

I mean the flag can stay in the funcbench tool as an option if someone needs to set it something different than the default, but why can't see skip setting it in the workflow file and leave it to some default option?

fewer flags to think about less mental baggage :smile:

Comment (krasi-georgiev) [reply prombench/manifests/funcbench/benchmark/2_job.yaml]: I see that the default is `/tmp/funcbench` so why do you need to set it to `/build` ?

File: funcbench/README.md
Comment (nevill) [reply funcbench/README.md]: Yes, we can remove this line. It just emphasizes that it's always executed. In real case, we will have `if: failure()` and `if: success()` behind. Using `if: always()` makes sure that the exit code of step `Run funcbench` can be passed down the chain (not interrupted by the exit code of step `Recycle all`).

**Review by geekodour — COMMENTED** [review]
@nevill can't we have this in the docker image? Or maybe just put this in the readme? Does not having this break something? Sorry i removed it without notice and forgot to mention! 

context : ~b28e84f~ 732eac7

**Review by krasi-georgiev — APPROVED** [review]

**Comment by nevill** [comment]
> @nevill can't we have this in the docker image? Or maybe just put this in the readme? Does not having this break something? Sorry i removed it without notice and forgot to mention!
> 
Because we don't have `gcc` in docker image, but CGO relies on it. It's introduced by [#275](https://github.com/prometheus/test-infra/pull/275/files#diff-d4ecb9c1cc7e2fbe0ec77c6deffa7eb2), and was mentioned in README but removed  [here](https://github.com/prometheus/test-infra/pull/346/files#diff-a164d67aaecfad99caff7e1c6c6b2357L58).

**Comment by nevill** [comment]
I removed `GO111MODULE=on` from `newGitHubEnv`, since our image is based go-1.13, and prometheus source is outside `GOPATH`.

**Comment by geekodour** [comment]
merged. Thanks!

**Pull request merged**
